src/AvifLoad.cs

```diff
@@ -141,5 +162,148 @@ private static void AddAvifMetadataToDocument(Document doc, AvifReader reader)
+            }
+        }
+
+        private static void SetOutputLayerDataRgba64(AvifReaderImage image, Surface output)
+        {
+            if (image.IsPremultipliedAlpha)
+            {
+                using (IBitmap asPrgba64 = image.Image.Cast<ColorPrgba64>())
+                using (IBitmapSource<ColorBgra32> converted = asPrgba64.CreateFormatConverter<ColorBgra32>())
+                {
+                    converted.CopyPixels(output.AsRegionPtr().Cast<ColorBgra32>());
+                }
+            }
+            else
+            {
+                using (IBitmapSource<ColorBgra32> converted = image.Image.CreateFormatConverter<ColorBgra32>())
+                {
+                    converted.CopyPixels(output.AsRegionPtr().Cast<ColorBgra32>());
+                }
+            }
+        }
+
+        private static void SetOutputLayerDataRgba128PQ(
+            Document document,
+            AvifReaderImage image,
+            Surface output,
+            IImagingFactory imagingFactory)
+        {
+            using (IColorContext dp3ColorContext = imagingFactory.CreateColorContext(KnownColorSpace.DisplayP3))
+            using (IDirect2DFactory d2dFactory = Direct2DFactory.Create())
+            {
+                if (image.IsPremultipliedAlpha)
+                {
+                    using (IBitmapSource<ColorPbgra32> dp3Image = PQToDisplayP3(image.Image,
+                                                                                ColorManagementAlphaMode.Premultiplied,
+                                                                                imagingFactory,
+                                                                                d2dFactory,
+                                                                                dp3ColorContext))
+                    {
+                        dp3Image.CopyPixels(output.AsRegionPtr().Cast<ColorPbgra32>());
+                    }
+                }
+                else
+                {
+                    // Direct2D requires everything to be "premultiplied"
+                    using (IBitmap asPrgba128Float = image.Image.Cast<ColorPrgba128Float>())
+                    using (IBitmapSource<ColorPbgra32> dp3Image = PQToDisplayP3(asPrgba128Float,
+                                                                                ColorManagementAlphaMode.Straight,
+                                                                                imagingFactory,
+                                                                                d2dFactory,
+                                                                                dp3ColorContext))
+                    {
+                        dp3Image.CopyPixels(output.AsRegionPtr().Cast<ColorPbgra32>());
+                    }
+                }
+
+                document.SetColorContext(dp3ColorContext);
+            }
+
+            static IBitmapSource<ColorPbgra32> PQToDisplayP3(
```

Comment:
Because this method takes the color context as a parameter, and is therefore not actually involved in the decision about whether that color context is DP3 or not, it might be useful to name it as such. Like, `PQToOtherColorContext`. Or you could create the DP3 color context inside the method and keep the name. (IMO)

Reply:
I had to create the D2D factory and color context outside of the method due to a use-after-free bug when calling CopyPixels.

Reply:
In that case you may need to create a second ref for use inside the delegate closure, and then dispose it at the end of the closure.

```cs
public static IBitmapSource Example(IDirect2DFactory factory, IColorContext colorContext)
{
    IColorContext colorContextP = colorContext.CreateRef();
    return factory.CreateBitmapSourceFromImage<ColorPbgra32>(
        ...
        delegate (IDeviceContext dc)
        {
            ...
            colorContextP.Dispose();
            return image;
        });
}
```

Otherwise it is certainly possible that the refs used inside the delegate will have already been disposed by the time the delegate itself is invoked.